### ISSUE TITLE
Reimplementation in RL4CO 

### ISSUE THREAD

**Issue by fedebotu** [wouterkool/attention-learn-to-route#58]
Hi there 👋🏼

First of all, thanks a lot for your library, it has inspired several works in our research group!
We are actively developing [RL4CO](https://github.com/kaist-silab/rl4co), a library for all things Reinforcement Learning for Combinatorial Optimization. We started the library by modularizing the Attention Model, which is the basis for several other autoregressive models. We also used some recent software (such as TorchRL, TensorDict, PyTorch Lightning and Hydra) as well as routines such as FlashAttention, and made everything as easy to use as possible in the hope of helping practitioners and researchers.

We welcome you to check RL4CO out ^^

**Comment by fedebotu** [comment]
Thanks for your quick answer 🚀 
- In terms of **performance**: yes! While reimplementing the AM we carefully checked whether everything was working step-by-step; in the exact same settings as the original code (same parameters, learning rate, batch sizes,  validation datasets...) we actually obtained slightly better performance:
![image](https://github.com/wouterkool/attention-learn-to-route/assets/48984123/b1522f05-07fa-441c-9070-98e990d3e828)
![image](https://github.com/wouterkool/attention-learn-to-route/assets/48984123/34472ccb-5db4-42b0-8100-5d116b401208)
with the only difference being the MHA module with the linear bias set to `true` -  the bump in the performance of the last 20 epochs is a step of `MultiStepLR` scheduler.
- In terms of **training time**: this depends on the implementation. To be honest, your code was already really well-optimized although it has been around for a few years! In theory, we should get similar / slightly better performance since we use [FlashAttention](https://github.com/Dao-AILab/flash-attention) in the encoder. In practice, however, we found that we are slightly slower with the same hyperparameters (on a single RTX 3090 we can train your original AM for TSP50 in 14 hours and ours in 15 hours). The reason is that we use [TensorDicts](https://github.com/pytorch/tensordict) and [TorchRL](https://github.com/pytorch/rl/tree/main/torchrl/envs) environments - we found the "culprits" are [data loading](https://github.com/pytorch/tensordict/issues/374) and most importantly the fact that we need to [re-create on the fly TensorDicts with large batch sizes _at each step_](https://github.com/kaist-silab/rl4co/issues/56), which has a pretty big impact on performance. One simple way to solve this is to just refactor environments in pure PyTorch. Right now [we are in contact with the TorchRL team](https://github.com/kaist-silab/rl4co/issues/72) and there have been several updates (just a few hours ago, the stable TensorDict 0.2.0 was released), so we plan to solve this in the next couple of weeks ;) There is another trick to noticeably improve performance. Setting `mask_inner` to `False` enables the FlashAttention routine during decoding (so for each step) and the training above takes 12.5 hours, even with the current slow TensorDict problem!  This of course, does degrade performance, but FlashAttention with masking may be [added](https://github.com/Dao-AILab/flash-attention/issues/352) in the near future, so it holds good promise!
- In terms of **evaluation time**: we found our evaluation script to be slightly faster - but if we check a single forward pass of the model, we get a similar trend as above, so we think there may have been some bottleneck in the eval script.

We would be more than happy to address your feedback if you check out RL4CO, you may contact us any time 😄

**Comment by fedebotu** [comment]
(Late) edit: now we are way more efficient as explained [here](https://github.com/ai4co/rl4co/releases/tag/v0.3.0)!

**Comment by wouterkool** [comment]
Great! I have added a link in the readme. However, I wonder if you have also had a look at https://github.com/cpwan/RLOR, they claim a 8x speedup over this repo using PPO.

**Comment by fedebotu** [comment]
Yes, we are aware of it!
From our understanding and our testing, their speedup is actually considered as the time to reach a certain cost as seen in Table 4. AM trains TSP 50 reaching 5.80 in 24 hours, while their PPO implementation - with some training and testing tricks - takes 3 hours. 
So, it is not a speedup per se - actually, due to the environment being in Numpy, even though vectorized, data collection is naturally a bottleneck - but rather, the time to reach a target performance. Besides, the comparison is made with their AM trained with PPO with larger batch size and learning rate and tested with the "multi-greedy" decoding scheme during inference (what in RL4CO we call multistart, i.e., the POMO decoding scheme that starts decoding from all nodes and then takes the best trajectory), while the baseline AM is just evaluated with one-shot greedy decoding. 
For these reasons, we think the 8x speedup claim is a bit overstated 👀
